### PR TITLE
[mle] implement alternate RLOC16 usage during role transition

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces a mechanism for a device transitioning from child to router role to keep receiving frames addressed to its previous short address for a brief period.

A new radio platform API, `otPlatRadioSetAlternateShortAddress()`, is introduced. This allows the OT stack to configure an alternate short address. Radio platform support for this function is indicated by the `OT_RADIO_CAPS_ALT_SHORT_ADDR` capability in `otPlatRadioGetCaps()` The same function can be used with `OT_RADIO_INVALID_SHORT_ADDR` (`0xfffe`) to clear a previously set alternate short address. Support for the new API is implemented in RCP and `RadioSpinel`, ensuring backward compatibility by dynamically checking supported radio capabilities.

MLE code is updated to instruct the radio to use the old child RLOC16 as an alternate address upon role transition. The MLE layer will automatically clear the alternate address after eight seconds, or if other state/mode changes occur. This eight-second window ensures the new router can transmit four MLE Advertisement messages.

----

- Related to [SPEC-1325](https://threadgroup.atlassian.net/browse/SPEC-1325)

### CODE REVIEW

**Review by LuDuda — APPROVED** [review]
All looks good! :+1:

How do you plan to test this feature?

**Comment by abtink** [comment]
> How do you plan to test this feature?

Thanks @LuDuda. I added a test that covers the basic behavior of this, i.e. the stack sets the "alt short address" after transition from child to router and after the timeout removes it.

**Comment by arnulfrupp** [comment]
This pull request has not been merged yet. I believe it would fix an issue critical to DALI+ certification tests. (See: https://github.com/orgs/openthread/discussions/9974).
Shall we make it a Threadgroup JIRA item item to make other vendors aware of the additional platform API needed. I assume this is the blocking issue !?

**Comment by abtink** [comment]
> Shall we make it a Threadgroup JIRA item item to make other vendors aware of the additional platform API needed. I assume this is the blocking issue !?

Yes. Maybe helpful to discuss this in a JIRA (though we initially assumed may not be necessary).

I rebased the PR on `main` and addressed the conflicts.

**Comment by George-Stefan** [comment]
> > Shall we make it a Threadgroup JIRA item item to make other vendors aware of the additional platform API needed. I assume this is the blocking issue !?
> 
> Yes. Maybe helpful to discuss this in a JIRA (though we initially assumed may not be necessary).
> 
> I rebased the PR on `main` and addressed the conflicts.
> 
> Edit: Created [SPEC-1325](https://threadgroup.atlassian.net/browse/SPEC-1325)

Hi @abtink,
I don't see where the new radio capability (OT_RADIO_CAPS_ALT_SHORT_ADDR) is checked for.
Everywhere it's assumed as present starting with the call to SetAlternateRloc16().

https://github.com/openthread/openthread/pull/10006/files#diff-00af97322ba69de16d755e6948f5866d9c58fd877353b184b19d98b52879dbceR3371

**Comment by abtink** [comment]
> Hi @abtink, I don't see where the new radio capability (`OT_RADIO_CAPS_ALT_SHORT_ADDR`) is checked for. Everywhere it's assumed as present, starting with the call to `SetAlternateRloc16()`, which ends up calling `otPlatRadioSetAlternateShortAddress()`.
>
> [https://github.com/openthread/openthread/pull/10006/files#diff-00af97322ba69de16d755e6948f5866d9c58fd877353b184b19d98b52879dbceR3371](https://github.com/openthread/openthread/pull/10006/files#diff-00af97322ba69de16d755e6948f5866d9c58fd877353b184b19d98b52879dbceR3371)
>
> Is it okay to silently fail?

It is used in `radio_spinel.cpp` to check whether the related spinel property is supported by RCP and can be used:

```c++
otError RadioSpinel::SetAlternateShortAddress(uint16_t aAddress)
{
    otError error = OT_ERROR_NONE;

    VerifyOrExit(sRadioCaps & OT_RADIO_CAPS_ALT_SHORT_ADDR);
    error = Set(SPINEL_PROP_MAC_15_4_ALT_SADDR, SPINEL_DATATYPE_UINT16_S, aAddress);

exit:
    return error;
}
```

Also, note that a weak (default) empty implementation of `otPlatRadioSetAlternateShortAddress()` is provided in `src/core/radio/radio_platform.cpp` (this is similar to other optional `otPlatRadio` APIs).

**Comment by George-Stefan** [comment]
> > Hi @abtink, I don't see where the new radio capability (`OT_RADIO_CAPS_ALT_SHORT_ADDR`) is checked for. Everywhere it's assumed as present, starting with the call to `SetAlternateRloc16()`, which ends up calling `otPlatRadioSetAlternateShortAddress()`.
> > https://github.com/openthread/openthread/pull/10006/files#diff-00af97322ba69de16d755e6948f5866d9c58fd877353b184b19d98b52879dbceR3371
> > Is it okay to silently fail?
> 
> @George-Stefan, it is used in `radio_spinel.cpp` to check whether the related spinel property is supported by RCP and can be used:
> 
> ```c++
> otError RadioSpinel::SetAlternateShortAddress(uint16_t aAddress)
> {
>     otError error = OT_ERROR_NONE;
> 
>     VerifyOrExit(sRadioCaps & OT_RADIO_CAPS_ALT_SHORT_ADDR);
>     error = Set(SPINEL_PROP_MAC_15_4_ALT_SADDR, SPINEL_DATATYPE_UINT16_S, aAddress);
> 
> exit:
>     return error;
> }
> ```
> 
> Also, note that a weak (default) empty implementation of `otPlatRadioSetAlternateShortAddress()` is provided in `src/core/radio/radio_platform.cpp` (this is similar to other optional `otPlatRadio` APIs).

Hi @abtink,
For RCP host build, VerifyOrExit(sRadioCaps & OT_RADIO_CAPS_ALT_SHORT_ADDR) doesn't change return status of  RadioSpinel::SetAlternateShortAddress() from OT_ERROR_NONE.

And this status is not propagated back to SetAlternateRloc16() (which doesn't return anything) and isn't checked in MleRouter::HandleAddressSolicitResponse().

For full SoC build, there is no check for OT_RADIO_CAPS_ALT_SHORT_ADDR adn there is no return status for the new added functions/methods

**Comment by abtink** [comment]
> So, is it okay for `SetAlternateRloc16()` to silently fail?
> From the implementation, I conclude that the mechanism is best effort/optional

Yes. Exactly. The OT stack will ask the radio to use an alternate address, but if the radio cannot (or does not yet support it), it is fine to ignore it; no further action (error handling) is needed in the OT core.

File: src/core/thread/mle_router.cpp [hunk 1]
```diff
@@ -88,6 +89,30 @@ MleRouter::MleRouter(Instance &aInstance)
...
 }
 
+void MleRouter::SetAlternateRloc16(uint16_t aRloc16)
+{
+    VerifyOrExit(aRloc16 != Mac::kShortAddrInvalid);
+
+    LogInfo("Setting alternate RLOC16 0x%04x", aRloc16);
+
+    Get<Mac::Mac>().SetAlternateShortAddress(aRloc16);
+    mAlternateRloc16Timeout = kAlteranteRloc16Timeout;
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp:99]:
```suggestion
    mAlternateRloc16Timeout = kAlternateRloc16Timeout;
```

File: src/core/thread/mle_router.hpp [hunk 2]
```diff
@@ -515,6 +515,7 @@ class MleRouter : public Mle
     static constexpr uint16_t kUnsolicitedDataResponseJitter = 500; // Max delay for unsol Data Response (in msec).
     static constexpr uint8_t  kLeaderDowngradeExtraDelay     = 10;  // Extra delay to downgrade leader (in sec).
     static constexpr uint8_t  kDefaultLeaderWeight           = 64;
+    static constexpr uint8_t  kAlteranteRloc16Timeout        = 8; // Time to use alternate RLOC16 (in sec).
```
Comment (jwhui) [inline src/core/thread/mle_router.hpp:518]:
```suggestion
    static constexpr uint8_t  kAlternateRloc16Timeout        = 8; // Time to use alternate RLOC16 (in sec).
```

**Pull request merged**